basic/payment-splitter/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+        }
+
+        /// Adds a shareholder to the PaymentSplitter
+        /// 
+        /// This is an authenticated method which needs to be called and presented the admin badge (obtained from creating the component).
+        /// This method is used to add a shareholder to the PaymentSplitter by minting an NFT to authenticate and keep track of the shareholder
+        /// data and then airdropping it to the specific shareholder that we specified the address for.
+        /// 
+        /// # Arguments
+        /// 
+        /// * `shareholder_address` - The address of the shareholder. This address will be airdropped the NFT.
+        /// * `shareholder_shares` - The amount of shares that belong to this shareholder.
+        #[auth(admin_badge_def, keep_auth)]
+        pub fn add_shareholder(
+            &mut self,
+            shareholder_address: Address,
+            shareholder_shares: Decimal
+        ) {
+            // Creating this NFT for this specific shareholder and storing it in a bucket
+            let nft_bucket: Bucket = self.shareholder_def.mint_nft(
+                self.number_of_shareholders,
+                ShareHolder {
+                    address: shareholder_address,
+                    shares: shareholder_shares,
+                    total_xrd_owed: Decimal(0),
+                    total_xrd_withdrawn: Decimal(0)
+                },
+                auth
+            );
+            self.number_of_shareholders += 1;
+
+            // Airdropping the NFT to the user with the specified address
+            Account::from(shareholder_address).deposit(nft_bucket);
+
+            // Logging the event
+            info!("Added shareholder {} with shares {} to the splitter. Number of current shareholders: {}", shareholder_address, shareholder_shares, self.number_of_shareholders)
+        }
+
+        /// Used to deposit XRD into the XRD vault controlled by this component.
+        /// 
+        /// This method is used to deposit XRD into the vault that this component controls. When this XRD is deposited into the Vault,
+        /// we update the NFT metadata to add XRD to the balance of the shareholders that they may withdraw at any point of time.
+        /// 
+        /// # Arguments
+        /// 
+        /// * `xrd_bucket` - A bucket of XRD that we're depositing into our component.
+        pub fn deposit_xrd(&mut self, xrd_bucket: Bucket) {
+            // Getting the amount of XRD that passed in the XRD bucket
+            let xrd_amount: Decimal = xrd_bucket.amount();
+            info!("Depositing XRD of amount: {}", xrd_amount);
+
+            // Taking the XRD bucket passed and depositing it into the XRD vault
+            self.xrd_vault.put(xrd_bucket);
+
+            // Getting the total number of shares that we have across all of our shareholders
+            let mut total_quantity_of_shares: Decimal = Decimal(0);
+            for i in 0..self.number_of_shareholders {
```

Comment:
Instead of doing that loop every time XRD is added, you could keep a field `total_quantity_of_shares` on the component state and update it inside the `add_shareholder` method